ksmt-symfpu/src/main/kotlin/org/ksmt/symfpu/solver/FpToBvTransformer.kt

```diff
@@ -0,0 +1,519 @@
+import org.ksmt.symfpu.operations.UnpackedFp.Companion.iteOp
+import org.ksmt.symfpu.operations.add
+import org.ksmt.symfpu.operations.bvToBool
+import org.ksmt.symfpu.operations.bvToFp
+import org.ksmt.symfpu.operations.divide
+import org.ksmt.symfpu.operations.equal
+import org.ksmt.symfpu.operations.fma
+import org.ksmt.symfpu.operations.fpToBv
+import org.ksmt.symfpu.operations.fpToFp
+import org.ksmt.symfpu.operations.greater
+import org.ksmt.symfpu.operations.greaterOrEqual
+import org.ksmt.symfpu.operations.isNegative
+import org.ksmt.symfpu.operations.isNormal
+import org.ksmt.symfpu.operations.isPositive
+import org.ksmt.symfpu.operations.isSubnormal
+import org.ksmt.symfpu.operations.less
+import org.ksmt.symfpu.operations.lessOrEqual
+import org.ksmt.symfpu.operations.max
+import org.ksmt.symfpu.operations.min
+import org.ksmt.symfpu.operations.multiply
+import org.ksmt.symfpu.operations.packToBv
+import org.ksmt.symfpu.operations.remainder
+import org.ksmt.symfpu.operations.roundToIntegral
+import org.ksmt.symfpu.operations.sqrt
+import org.ksmt.symfpu.operations.sub
+import org.ksmt.symfpu.operations.unpack
+import org.ksmt.symfpu.solver.ArraysTransform.Companion.packToBvIfUnpacked
+import org.ksmt.symfpu.solver.ArraysTransform.Companion.transformSortRemoveFP
+import org.ksmt.symfpu.solver.ArraysTransform.Companion.transformedArraySort
+import org.ksmt.symfpu.solver.SymFPUModel.Companion.declContainsFp
+import org.ksmt.utils.asExpr
+import org.ksmt.utils.cast
+
+class FpToBvTransformer(ctx: KContext) : KNonRecursiveTransformer(ctx) {
+    private val mapFpToBvDeclImpl = mutableMapOf<KDecl<*>, KConst<*>>()
+    val mapFpToBvDecl: Map<KDecl<*>, KConst<*>> get() = mapFpToBvDeclImpl
+
+    private val arraysTransform = ArraysTransform(ctx)
+
+    private val mapFpToUnpackedFpImpl =
+        mutableMapOf<KDecl<KFpSort>, UnpackedFp<KFpSort>>()
+
+    // for tests
+    val mapFpToUnpackedFp: Map<KDecl<KFpSort>, UnpackedFp<KFpSort>> get() = mapFpToUnpackedFpImpl
+
+
+    // use this function instead of apply as it may return UnpackedFp wrapper
+
+    fun <T : KSort> applyAndGetExpr(expr: KExpr<T>): KExpr<T> {
+        val applied = apply(expr)
+        // it might have UnpackedFp inside, so
+        // transform them to bvs
+        return AdapterTermsRewriter().apply(applied)
+    }
+
+    inner class AdapterTermsRewriter : KNonRecursiveTransformer(ctx) {
+        fun <T : KFpSort> transform(expr: UnpackedFp<T>): KExpr<KBvSort> = with(ctx) {
+            return packToBv(expr)
+        }
+    }
+
+
+    override fun <T : KSort> transform(expr: KEqExpr<T>): KExpr<KBoolSort> = with(ctx) {
+        transformExprAfterTransformed(expr, expr.lhs, expr.rhs) { l, r ->
+            if (l is UnpackedFp<*> && r is UnpackedFp<*>) {
+                val flags = mkAnd(l.isNaN eq r.isNaN, l.isInf eq r.isInf, l.isZero eq r.isZero)
+                if (l.packedBv is UnpackedFp.PackedFp.Exists && r.packedBv is UnpackedFp.PackedFp.Exists)
```

Comment:
Let's keep code style consistent across ksmt modules. We use braces for all `if` statements.